perf: Remove redundant copying of batches after FilterExec

andygrove opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes https://github.com/apache/datafusion-comet/issues/757

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

We were often creating a filtered batch in `FilterExec` and then making a copy of that batch in `CopyExec`, resulting in redundant copying of data in the case where `FilterExec` had already created a new batch with the filtered data. It was necessary to have the `CopyExec` because `FilterExec` can pass through input batches in the case where the predicate evaluates to true for all rows and this is not safe in Comet because we re-use arrays in the `ScanExec`.

This PR introduces a customized version of `FilterExec` that uses Arrow's `take_record_batch` kernel to always create new batches, even in the case where the predicate evaluates to true for all rows. This removes the need to wrap `FilterExec` in a `CopyExec`.

This reduces our TPC-DS time by ~25 seconds.

![tpcds_allqueries](https://github.com/user-attachments/assets/91df5a61-57e3-4ef2-adfe-259a9d1b2a2d)

The majority of queries are faster with this change:

![tpcds_queries_speedup](https://github.com/user-attachments/assets/159a306e-2a89-4ac1-9b1d-66728cd050ad)

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

- Copy DataFusion's FilterExec into Comet and make a one line change so that FilterExec never passes through input batches -- I will suggest some changes upstream so that we can customize rather than duplicate
- Stop wrapping FilterExec in CopyExec

## How are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

- Existing tests
- I ran TPC-DS without error

andygrove commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,442 @@
+        })
+}
+
+// BEGIN Comet changes
+pub fn filter_record_batch(
+    record_batch: &RecordBatch,
+    predicate: &BooleanArray,
+) -> std::result::Result<RecordBatch, ArrowError> {
+    // turn predicate into selection vector
+    let mut sv = Int32Builder::with_capacity(predicate.true_count());
+    for i in 0..predicate.len() {
+        if !predicate.is_null(i) && predicate.value(i) {
+            sv.append_value(i as i32);
+        }
+    }
+    let sv = sv.finish();
+    // note that this does not unpack dictionary-encoded arrays
+    take_record_batch(record_batch, &sv)
+}
+// END Comet changes
```
This is the modification to FilterExec

andygrove commented:
@mbutrovich you may be interested in reviewing this PR

Dandandan commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,442 @@
+    predicate: &BooleanArray,
+) -> std::result::Result<RecordBatch, ArrowError> {
+    // turn predicate into selection vector
+    let mut sv = Int32Builder::with_capacity(predicate.true_count());
```
Probably special casing the edge case on `true_count` == `num_rows` and defaulting to `filter` otherwise would avoid regressing "normal filtering"

Dandandan commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,442 @@
+    }
+    let sv = sv.finish();
+    // note that this does not unpack dictionary-encoded arrays
+    take_record_batch(record_batch, &sv)
```
Probably a better way to copy is to use MutableArrayData, like done in concatenate:

https://docs.rs/arrow-select/52.2.0/src/arrow_select/concat.rs.html#180

Dandandan replied:
Ah, this is done in`CopyExec` already - this should be faster.

andygrove replied:
The goal of this PR is to no longer need to use CopyExec after a FilterExec (to avoid copying twice in some cases)

Dandandan replied:
Yes got it. It might be possible though to reuse the same code from copyexec in here for copying the recordbatch?

mbutrovich commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,442 @@
+    // turn predicate into selection vector
+    let mut sv = Int32Builder::with_capacity(predicate.true_count());
+    for i in 0..predicate.len() {
+        if !predicate.is_null(i) && predicate.value(i) {
```
Another possible special case that we should measure the impact of: we could elide `!predicate.is_null(i)` if we know there are no nulls in the predicate array. You end up with duplicated loops (one with and without the null check) but depending on the branch predictor this may be worth it.

andygrove commented:
Thanks @Dandandan @viirya for the feedback so far. I have now added criterion benchmarks before I start addressing the feedback. Here are the current results.

```
filter/arrow_filter_record_batch - few
                        time:   [15.546 µs 15.601 µs 15.682 µs]
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe
filter/arrow_filter_record_batch - many
                        time:   [74.935 µs 75.129 µs 75.324 µs]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
filter/arrow_filter_record_batch - all
                        time:   [396.57 ns 397.80 ns 399.05 ns]
filter/comet_filter - few
                        time:   [21.772 µs 22.012 µs 22.337 µs]
Found 14 outliers among 100 measurements (14.00%)
  2 (2.00%) high mild
  12 (12.00%) high severe
Benchmarking filter/comet_filter - many: Warming up for 500.00 ms
Warning: Unable to complete 100 samples in 500.0ms. You may wish to increase target time to 762.3ms, enable flat sampling, or reduce sample count to 50.
filter/comet_filter - many
                        time:   [146.44 µs 147.25 µs 148.01 µs]
Found 12 outliers among 100 measurements (12.00%)
  12 (12.00%) low mild
Benchmarking filter/comet_filter - all: Warming up for 500.00 ms
Warning: Unable to complete 100 samples in 500.0ms. You may wish to increase target time to 895.1ms, enable flat sampling, or reduce sample count to 50.
filter/comet_filter - all
                        time:   [177.25 µs 183.04 µs 190.27 µs]
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe
```

viirya commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,450 @@
+}
+
+// BEGIN Comet changes
+pub fn comet_filter_record_batch(
```
```suggestion
fn comet_filter_record_batch(
```

andygrove replied:
I think it has to be public so that we can access it from the criterion benchmark but I will double check this.

viirya replied:
Okay.

andygrove commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,442 @@
+    }
+    let sv = sv.finish();
+    // note that this does not unpack dictionary-encoded arrays
+    take_record_batch(record_batch, &sv)
```
That is correct. The goal is to ensure that FilterExec *always* copies instead of only copying most of the time, so that we avoid also having a CopyExec that performs yet another copy.

I have implemented the suggestion from @Dandandan to specialize for the case where all rows are selected

viirya commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,450 @@
+        })
+}
+
+// BEGIN Comet changes
```
```suggestion
// BEGIN Comet changes
// `FilterExec` could modify input batch or return input batch without change. Instead of always
// adding `CopyExec` on top of it, we only copy input batch for the special case.
```

andygrove commented:
After addressing the first round of feedback, we now have:

```rust
pub fn comet_filter_record_batch(
    record_batch: &RecordBatch,
    predicate: &BooleanArray,
) -> std::result::Result<RecordBatch, ArrowError> {
    if predicate.true_count() == record_batch.num_rows() {
        // special case where we just make an exact copy
        let arrays: Vec<ArrayRef> = record_batch
            .columns()
            .iter()
            .map(|array| {
                let capacity = array.len();
                let data = array.to_data();
                let mut mutable = MutableArrayData::new(vec![&data], false, capacity);
                mutable.extend(0, 0, capacity);
                make_array(mutable.freeze())
            })
            .collect();
        let options = RecordBatchOptions::new().with_row_count(Some(record_batch.num_rows()));
        RecordBatch::try_new_with_options(record_batch.schema().clone(), arrays, &options)
    } else {
        filter_record_batch(record_batch, predicate)
    }
}
```

New benchmark results:

```
filter/comet_filter - few
                        time:   [14.650 µs 14.727 µs 14.831 µs]
                        change: [-36.702% -35.128% -33.681%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
filter/comet_filter - many
                        time:   [75.962 µs 76.172 µs 76.381 µs]
                        change: [-48.681% -48.501% -48.303%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
filter/comet_filter - all
                        time:   [34.497 µs 34.628 µs 34.764 µs]
                        change: [-80.854% -80.527% -80.256%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
```

This certainly looks a lot better. I am running TPC-DS again to make sure this really is always copying. I had tried an approach like this in the past but ran into data corruption issues.

viirya commented on native/Cargo.toml:
```diff
@@ -44,6 +44,7 @@ datafusion = { default-features = false, git = "https://github.com/apache/datafu
 datafusion-functions = { git = "https://github.com/apache/datafusion.git", rev = "41.0.0-rc1", features = ["crypto_expressions"] }
 datafusion-functions-nested = { git = "https://github.com/apache/datafusion.git",  rev = "41.0.0-rc1", default-features = false }
 datafusion-expr = { git = "https://github.com/apache/datafusion.git", rev = "41.0.0-rc1", default-features = false }
+datafusion-execution = { git = "https://github.com/apache/datafusion.git", rev = "41.0.0-rc1", default-features = false }
```
We need `datafusion-execution`?

andygrove replied:
We use `use datafusion::execution::TaskContext`. I guess we were just pulling this in transitively before via the `datafusion` crate rather than being explicit. 

We may want to avoid bringing in the core `datafusion` crate and just depend directly on the crates that we need.

andygrove commented:
The results are even better now :tada: 

![tpcds_allqueries](https://github.com/user-attachments/assets/c45189c1-ec94-4fd0-a323-fbbcc00bf672)

![tpcds_queries_speedup](https://github.com/user-attachments/assets/e9e8f603-f597-49a5-873b-ecc38d0ba2d9)

mbutrovich commented on native/core/src/execution/operators/filter.rs:
```diff
@@ -0,0 +1,452 @@
+            .map(|array| {
+                let capacity = array.len();
+                let data = array.to_data();
+                let mut mutable = MutableArrayData::new(vec![&data], false, capacity);
```
More of a Rust question for myself to dig into: is there a reason we can't do `array.to_data().clone()` and then call `make_array` on that? It seems inefficient to allocate a resizable data structure, truncate it with `extend()` then `freeze()` it, but I am still very new to this. Thankfully there are good benchmarks in this PR for me to explore that on my own. :)

viirya replied:
`clone` will have a shallow copy with the buffers are still shared, but we need actually copying the data.

mbutrovich replied:
I thought `copy()` would make a shallow copy, while `clone()` would copy the data?

viirya replied:
The buffer implementation under the arrays holds references. `clone` will just copy the references without copying data.

mbutrovich replied:
So `RecordBatch` stores a `Vec` of reference counted arrays `columns: Vec<Arc<dyn Array>>`? That makes sense to me that if we `clone()` we just bump ref counts.  But the underlying `ArrayData` should be safe to `clone()` because its underlying `Vec` will deep copy its buffer.

mbutrovich replied:
Ah, I see the reference counted stuff all the way down inside `Buffer`. I understand why it won't work now. Thanks!

viirya replied:
This is the snippet of `Buffer` implementation in arrow-rs:

```rust
pub struct Buffer {
    /// the internal byte buffer.
    data: Arc<Bytes>,
    ...
}
```


And `ArrayData`:

```rust
pub struct ArrayData {
    ...
    buffers: Vec<Buffer>,
    ...
}
```

When you clone it, `Buffer` won't copy data for you but you only get a clone of `Arc`.

andygrove commented:
I created a PR against DataFusion to add a flag to FilterExec so that we can switch back to using the upstream version, assumign this PR gets accepted.

andygrove commented:
Thanks for the reviews @viirya @Dandandan @mbutrovich
